### PR TITLE
CircleCI Migration

### PR DESCRIPTION
**Pull request by sabre1041**
Migrates existing workflow on CircleCI job to GitHub actions native build

* Creation of release workflow
* Updates [build-pr job](https://github.com/helm/helm/blob/main/.github/workflows/build-pr.yml) to incorporate building and uploading binaries
* Updates [code-analysis job](https://github.com/helm/helm/blob/main/.github/workflows/codeql-analysis.yml) to pin dependencies

### Tasks:

- [ ] Add `AZURE_STORAGE_CONNECTION_STRING` and `AZURE_STORAGE_CONTAINER_NAME` as GitHub Secrets
- [ ] Ensure all CircleCI assets removed

Builds upon #9989

### CODE REVIEW

File: .github/workflows/release.yml
Comment (mattfarina) [inline .github/workflows/release.yml:23]: Since this is already filtered for tags, is there a reason we aren't using the `GITHUB_REF_NAME` variable?

File: .github/workflows/release.yml
Comment (mattfarina) [inline .github/workflows/release.yml:1]: Should this PR remove the deploy from CircleCI? So that the next release does not have two systems pushing it.

File: .github/workflows/build-pr.yml
Comment (sabre1041) [reply .github/workflows/build-pr.yml:36]: Everything CircleCI should be removed. I can certainly add the removal of the circleci script if we would like in this PR.

File: .github/workflows/release.yml
Comment (sabre1041) [reply .github/workflows/release.yml:1]: Everything CircleCI should be removed. I can certainly add the removal of the circleci script if we would like in this PR.

File: .github/workflows/release.yml
Comment (sabre1041) [reply .github/workflows/release.yml]: Fixed

File: .github/workflows/build-pr.yml
Comment (sabre1041) [reply .github/workflows/build-pr.yml:59]: done

**Review by scottrigby — COMMENTED** [review]
Hey @sabre1041 👋 I closed #11820 in favor of this one. I created that PR only because I thought you said this work had yet to be finished and I got frustrated with Circle again so just busted it out. I see now that you meant your PR just needed to be finished. Kind of good to see though that the solution is roughly the same in both PRs.

As I said in https://github.com/helm/helm/pull/11820#issuecomment-1433346734:

> I think there are some things in my PR that would improve yours, but your PR has good things like pinning by sha, and it was first. So let's work on it together from yours and get this in 😄 

Here's what 'd like to see merged from the two:

- Pinning by sha:
    Nice improvement overall (for everyone else, this is rec'd by GitHub [actions hardening doc](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions))
- adding `make test build` from Circle config:
    Both PRs add this because it was missing, and this is clearly the logical place 👍 
- adding logic from deploy.sh to Circle config:
    - I added them both to a single release workflow, just as they were together in the script
    - You added the canary ("main" branch) release to build-pr.yml, and tagged releases to a new release.yml. But I don't think canary releases belong in the PR workflow (honestly that should just be renamed "test", but either way it is just for PRs). I also think there's value in the release to AZ logic in the same file, just so that practically when improvements are made they're done in the same place. The only difference between the two in the script is a single `--overwrite` flag, which I noted in my PR
- Very minor. It's not necessary to export `VERSION` from a previous build step

File: .github/workflows/build-pr.yml
Comment (lucasrod16) [inline .github/workflows/build-pr.yml:18]: This is more of a question/suggestion than a "we should do this". Curious if there would be any value in using the `go-version-file` input with the setup-go action. It takes the path to the `go.mod` file, and uses the version specified there, so when the go version is updated, it would automatically be reflected in the workflows. [Link](https://github.com/actions/setup-go#getting-go-version-from-the-gomod-file) to the docs. My thinking is that it would be one less place to worry about updates/maintenance, but I'm curious to hear others' thoughts.

Comment (sabre1041) [reply .github/workflows/build-pr.yml:18]: thats a new(er) enhancement and I like that it will reduce the number of changes that need to be made when adopting new versions of go.

Thanks for calling it out!

Comment (scottrigby) [reply .github/workflows/build-pr.yml]: @lucasrod16 I didn't know about this feature. Very nice!

Comment (scottrigby) [reply .github/workflows/build-pr.yml]: @lucasrod16 for now, we have to revert back to specifying the version explicitly. Because it appears there was a regression (mismatch between what's actually in go.mod and the circle tests), that was introduced here 

https://github.com/helm/helm/blob/017785a2f1104eb4299f8d31feb550c213350f13/go.mod#L3

in this commit https://github.com/helm/helm/pull/10912/commits/ffa19a4b5d836283a91a4c16f8b81e734a973afc (see line https://github.com/helm/helm/pull/10912/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6) from this PR #10912. Note the commit comment:

> fix circleci tests

We revert go.mod back to 1.18 here but that would be scope creep for this PR. Let's do that in a follow-up PR, at which point we could use the `go-version-file` option.

File: .github/workflows/build-pr.yml
Comment (scottrigby) [inline .github/workflows/build-pr.yml:18]: See https://github.com/helm/helm/pull/11676#discussion_r1112595497

```suggestion
          go-version: '1.18'
```

File: .github/workflows/build-pr.yml
Comment (scottrigby) [inline .github/workflows/build-pr.yml:47]: See above

```suggestion
          go-version: '1.18'
```

File: .github/workflows/release.yml
Comment (scottrigby) [inline .github/workflows/release.yml:18]: See above

```suggestion
          go-version: '1.18'
```

File: .github/workflows/build-pr.yml
Comment (scottrigby) [reply .github/workflows/build-pr.yml]: Made follow-up issue https://github.com/helm/helm/issues/11838

**Review by scottrigby — CHANGES_REQUESTED** [review]
Change back to explicitly setting Go version. See https://github.com/helm/helm/pull/11676#discussion_r1112595497 and follow-up issue https://github.com/helm/helm/issues/11838.

File: .github/workflows/build-pr.yml
Comment (lucasrod16) [reply .github/workflows/build-pr.yml]: @scottrigby ahh I see, thanks for the context! That sounds good, I can write-up an issue and link this discussion for more context.

@sabre1041 apologies for the side-track here

Comment (lucasrod16) [reply .github/workflows/build-pr.yml]: @scottrigby nevermind I see that you've created a follow-up [issue](https://github.com/helm/helm/issues/11838) already.

**Review by scottrigby — APPROVED** [review]
Actions (on PRs) pass, and all actions lgtm! Our plan is to test the build/upload by both merging this PR into main and then cutting a dev tag to test the process for canary.

@sabre1041 I made one more code suggestion (wrapping up my comments above), as a minor speed/simplification improvement. Feel free to merge that suggestion, but either way adding my approval here now. Let's unblock the things 💖

File: README.md
Comment (scottrigby) [inline README.md:3]: I think we want the release badge for the Readme, right? According to GH UI, I believe this is the right syntax (if no branch query is specified, the default branch - main - is assumed)

```suggestion
[![Build Status](https://github.com/helm/helm/workflows/release.yml/badge.svg)](https://github.com/helm/helm/actions/workflows/release.yml)
```

**Review by scottrigby — COMMENTED** [review]
Also this one. Please check using another workflow name (as this wf does not yet exist to test)

**Comment by scottrigby** [comment]
Excellent! I think we should wait for one other maintainer to approve since you and I both have commits on this

**Comment by sabre1041** [comment]
@mattfarina any chance you might be able to review once again?

**Review by mattfarina — CHANGES_REQUESTED** [review]

**Review by mattfarina — APPROVED** [review]

**Comment by scottrigby** [comment]
🎉 🙌

**Pull request merged**
